### PR TITLE
Support custom hardware key prompt

### PR DESCRIPTION
**Pull request by gzdunek**
This PR adds an ability to pass a custom hardware key prompt to functions that talk to YubiKey.
`tsh` now uses a `cliPrompt`; Teleport Connect will have its own prompt, shown as a modal in the app.

The prompt interface is defined in `yubikey_common.go`. I extracted four methods from `yubikey.go`: `AskPIN`, `Touch`, `ChangePIN` and `ConformSlotOverwrite`. When the prompt message is generated dynamically (like in `ConformSlotOverwrite`) it can be passed to the method. Otherwise, the implementer has to provide the message. 

Converting most of the prompts to work with the new interface was fairly straightforward. The only significant change is `ChangePIN`. Previously, it set PUK directly after prompting for it, now I have to gather all data upfront (PIN & PUK), and then call the yubkiey functions.
Because of that, I check on the implementer side if the provided PUK is empty or default, ask for it, and then set `PINAndPUK.PUKChanged` flag so that I know if the PUK on the key should be changed.

### CODE REVIEW

File: api/utils/keys/privatekey.go [hunk 1]
```diff
@@ -214,14 +214,20 @@ func LoadPrivateKey(keyFile string) (*PrivateKey, error) {
 
 // ParsePrivateKey returns the PrivateKey for the given key PEM block.
 func ParsePrivateKey(keyPEM []byte) (*PrivateKey, error) {
+	return ParsePrivateKeyWithCustomPrompt(keyPEM, nil)
+}
+
+// ParsePrivateKeyWithCustomPrompt returns the PrivateKey for the given key PEM block.
+// Allows passing a custom hardware key prompt.
+func ParsePrivateKeyWithCustomPrompt(keyPEM []byte, customPrompt HardwareKeyPrompt) (*PrivateKey, error) {
```
Comment (gzdunek) [inline api/utils/keys/privatekey.go:222]: I don't really like this duality :(
Unfortunately, `ParsePrivateKey` is used in so many places, that updating all of them is too cumbersome. Additionally, I don't want to force the callsites to pass `nil` to this function when they don't really care about the `customPrompt`.

File: api/utils/keys/privatekey.go [hunk 2]
```diff
@@ -214,14 +214,20 @@ func LoadPrivateKey(keyFile string) (*PrivateKey, error) {
 
 // ParsePrivateKey returns the PrivateKey for the given key PEM block.
 func ParsePrivateKey(keyPEM []byte) (*PrivateKey, error) {
+	return ParsePrivateKeyWithCustomPrompt(keyPEM, nil)
```
Comment (Joerger) [inline api/utils/keys/privatekey.go:217]: nit:
```suggestion
	return ParsePrivateKeyWithCustomPrompt(keyPEM, nil /*customPrompt*/)
```

File: lib/client/keystore.go [hunk 3]
```diff
@@ -174,6 +182,10 @@ func (fs *FSKeyStore) kubeCredPath(idx KeyRingIndex, kubename string) string {
 	return keypaths.KubeCredPath(fs.KeyDir, idx.ProxyHost, idx.Username, idx.ClusterName, kubename)
 }
 
+func (fs *FSKeyStore) SetCustomHardwareKeyPrompt(prompt keys.HardwareKeyPrompt) {
```
Comment (Joerger) [inline lib/client/keystore.go:185]: nit: add godoc

File: lib/client/keystore.go [hunk 4]
```diff
@@ -865,3 +877,5 @@ func (ms *MemKeyStore) GetSSHCertificates(proxyHost, username string) ([]*ssh.Ce
 
 	return sshCerts, nil
 }
+
+func (ms *MemKeyStore) SetCustomHardwareKeyPrompt(_ keys.HardwareKeyPrompt) {}
```
Comment (Joerger) [inline lib/client/keystore.go:881]: nit: add godoc

File: api/utils/keys/yubikey_common.go [hunk 5]
```diff
@@ -19,15 +19,37 @@ import (
...
+	// If the PUK has a default value, it should ask for the new value for it.
+	// It is up to the implementer how the validation is handled.
+	// For example, CLI prompt can ask for a valid PIN/PUK in a loop, a GUI
+	// prompt can use the frontend validation.
+	ChangePIN(ctx context.Context) (*PINAndPUK, error)
+	// ConfirmSlotOverwrite asks the user if the slot's private key and certificate can be overridden.
+	ConfirmSlotOverwrite(ctx context.Context, message string) (bool, error)
+}
+
+type PINAndPUK struct {
```
Comment (Joerger) [inline api/utils/keys/yubikey_common.go:38]: nit: Add godocs

**Review by Joerger — COMMENTED** [review]
Nice add! LGTM just some minor nits

File: api/utils/keys/privatekey.go (hunk 1 above)
Comment (gzdunek) [reply api/utils/keys/privatekey.go]: This is really cool, applied now :D

**Review by Joerger — APPROVED** [review]

**Review by nklaassen — APPROVED** [review]

**Pull request merged**